src/ObjCRuntime/DisposableObject.cs

```diff
@@ -0,0 +1,95 @@
+//
+// DisposableObject.cs: A base class for for many native data types without assuming any particular lifecycle.
+
+// Authors:
+//   Rolf Bjarne Kvinge
+//
+// Copyright 2021 Microsoft Corp
+//
+using System;
+using System.Runtime.InteropServices;
+using ObjCRuntime;
+using Foundation;
+
+#nullable enable
+
+namespace ObjCRuntime {
+	//
+	// The DisposableObject class is intended to be a base class for many native data
+	// data types, without assuming any particular lifecycle (might be reference counted,
+	// might not be).
+	//
+	// It provides the common boilerplate for this kind of objects and the Dispose
+	// pattern.
+	//
+	public abstract class DisposableObject : INativeObject, IDisposable {
+		IntPtr handle;
+		readonly bool owns;
+
+		public IntPtr Handle {
+			get => handle;
+			protected set => InitializeHandle (value);
+		}
+
+		protected bool Owns { get; }
+
+		protected DisposableObject ()
+		{
+		}
+
+		protected DisposableObject (IntPtr handle, bool owns)
+			: this (handle, owns, true)
+		{
+		}
+
+		protected DisposableObject (IntPtr handle, bool owns, bool verify)
+		{
+			InitializeHandle (handle, verify);
+			this.owns = owns;
```

Comment:
Shoudln't this SuppressFinalize if the object is not owned?

Reply:
That's an interesting optimization! But what happens if a subclass override Dispose to free native resources independent on whether the handle is owned or not?

Reply:
Could be documented as a remarks, because I don't believe that is generally the case :p

Reply:
Removing the GC.SuppressFinalize is an easier change to allow for more cases than adding it in the futur breaking compat :D

Reply:
Actually it won't work: the NativeObject subclass will retain if `owns` is false - thus taking ownership - and it must still be collected.

Reply:
Doesn't that mean NativeObject should ReRegisterFinalize because it retains - thus owns?

Reply:
I think I'm missing some context here. If the ownership is taken by managed, then wouldn't `owns` be passed as true?

Reply:
> Doesn't that mean NativeObject should ReRegisterFinalize because it retains - thus owns?

Yes, I guess something like that could work. That means that in the end `NativeObject` instances would never suppress their finalize method, it would only have any effect for other `DisposableObject` subclasses.

> I think I'm missing some context here. If the ownership is taken by managed, then wouldn't `owns` be passed as true?

The `owns` parameter has two slightly different meaning depending on the type:

Passing `owns=false` means:

* For `NativeObject` subclasses it means that the class must retain the handle. At that point the `Disposable.Owns` field is more of a historical artifact, since it's not used (we could set it to `true` if that makes more sense).
* For other `DisposableObject` subclasses it currently means that the instance does not own the Handle, and must not free it.

Passing `owns=true` means that the caller has ownership of the handle in question, and gives it to the `DisposableObject` instance. The Handle must be freed in Dispose; `NativeObject` will release it, other subclasses will have their own logic.

The main difference between `NativeObject` and other `DisposableObject` subclasses is that the latter represents handles that can't be retained/released.